src/xsk/socket.rs

```diff
@@ -54,4 +47,53 @@ impl SocketFd {
         }
         Ok(SocketFd(fd))
     }
+
+    /// Get an option for the socket referred to by the current file descriptor, storing the result
+    /// in `val`
+    ///
+    /// Returns the actual size of the option's value
+    pub(crate) fn get_opt<T>(self, level: i32, name: i32, mut val: &T) -> Result<u32, Errno>
+    where
+        T: Sized,
+    {
+        let mut len: libc::socklen_t = core::mem::size_of_val(val) as libc::socklen_t;
+        let err = unsafe {
+            libc::getsockopt(
+                self.0,
+                level,
+                name,
+                (&mut val) as *mut _ as *mut libc::c_void,
+                &mut len,
+            )
+        };
+
+        if err != 0 {
+            return Err(LastErrno)?;
+        }
+
+        Ok(len)
+    }
```

Comment:
Seems buggy, we're letting the kernel mutate the reference/pointer? and communicating that the pointer has the size of `T` doesn't make any sense. Sockopt syscalls can not really be abstracted into a safe interface like this as they are all essentially different systemcalls depending on the level and name and the type of the buffer being passed matters a lot, i.e. it can not be independently generic like this.

Please avoid changing large portions any unsafe code if you want this PR to progress quickly.